poseidon2-starky/src/stark.rs

```diff
@@ -1 +1,285 @@
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = [P::ZEROS; 8];
+
+    for i in 0..8 {
+        for j in 0..4 {
+            out[i] = if i < 4 {
+                out[i] + state[j] * FE::from_canonical_usize(M4[i][j])
+            } else {
+                out[i] + state[j + 4] * FE::from_canonical_usize(M4[i - 4][j])
+            }
+        }
+    }
+
+    let mut stored = [P::ZEROS; 4];
+    for i in 0..4 {
+        stored[i] = out[i] + out[4 + i];
+    }
+    for i in 0..8 {
+        out[i] = out[i] + stored[i % 4];
+    }
+
+    out
+}
+
+// degree: 1
+fn matmul_internal8_constraints<
+    F: RichField + Extendable<D>,
+    const D: usize,
+    FE,
+    P,
+    const D2: usize,
+>(
+    state: &[P; 8],
+) -> [P; 8]
+where
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = [P::ZEROS; 8];
+    let mut sum = P::ZEROS;
+
+    for i in 0..STATE_SIZE {
+        sum = sum + state[i];
+    }
+
+    for i in 0..STATE_SIZE {
+        out[i] = state[i] * scalar_to_fe::<F, D, FE, D2, FpGoldiLocks>(MAT_DIAG8_M_1[i]);
+        out[i] = out[i] + sum;
+    }
+
+    out
+}
+
+// degree: 1
+fn add_rc_constraints<F: RichField + Extendable<D>, const D: usize, FE, P, const D2: usize>(
+    state: &[P; 8],
+    r: usize,
+) -> [P; 8]
+where
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = [P::ZEROS; 8];
+
+    for i in 0..8 {
+        out[i] = state[i] + scalar_to_fe::<F, D, FE, D2, FpGoldiLocks>(RC8[r][i]);
+    }
+
+    out
+}
+
+// degree: SBOX_DEGREE (7)
+fn sbox_p_constraints<F: RichField + Extendable<D>, const D: usize, FE, P, const D2: usize>(
+    state: &P,
+) -> P
+where
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = P::ONES;
+
+    for _ in 0..SBOX_DEGREE {
+        out = out.mul(*state);
+    }
+
+    out
+}
+
+#[derive(Copy, Clone, Default)]
+pub struct Poseidon2Stark<F, const D: usize> {
+    pub _f: PhantomData<F>,
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for Poseidon2Stark<F, D> {
+    const COLUMNS: usize = NUM_COLS;
+    const PUBLIC_INPUTS: usize = 0;
+
+    fn eval_packed_generic<FE, P, const D2: usize>(
+        &self,
+        vars: StarkEvaluationVars<FE, P, { Self::COLUMNS }, { Self::PUBLIC_INPUTS }>,
+        yield_constr: &mut ConstraintConsumer<P>,
+    ) where
+        FE: FieldExtension<D2, BaseField = F>,
+        P: PackedField<Scalar = FE>,
+    {
+        let lv = vars.local_values;
+        let mut state = matmul_external8_constraints(lv[0..8].try_into().unwrap());
+
+        // first full rounds
+        for r in 0..ROUNDS_F {
+            state = add_rc_constraints(&state, r);
+            for i in 0..STATE_SIZE {
+                state[i] = sbox_p_constraints(&state[i]);
+            }
+            state = matmul_external8_constraints(&state);
```

Comment:
here in matmul_external8..() following loop
```rust
   for i in 0..STATE_SIZE {
        out[i] = state[i] * scalar_to_fe::<F, D, FE, D2, FpGoldiLocks>(MAT_DIAG8_M_1[i]);
        out[i] = out[i] + sum;
    
  ```
    multiplies a term with each state element. Doesn't that increase degree to 8?

Reply:
It does not multiplies a term with each state element.
sum = s0 + s1 + ... + s7
out0 = s0 * const + sum
out1 = s1 * const + sum
...